### PR TITLE
Initial support for SaltStack.

### PR DESCRIPTION
**Pull request by geomacy**
See https://issues.apache.org/jira/browse/BROOKLYN-217.

This commit merges code from incubator-brooklyn, following the move to the new repos
after graduation. As such, it unfortunately loses the commit history, but the code
is based on the work done by Hadrian Zbarcea.  See the Jira link above for history
and discussion.

### CODE REVIEW

File: utils/common/src/main/java/org/apache/brooklyn/util/yaml/Yamls.java [hunk 1]
```diff
@@ -151,14 +151,14 @@ public static void dump(int depth, Object r) {
 
     /** simplifies new Yaml().loadAll, and converts to list to prevent single-use iterable bug in yaml */
     @SuppressWarnings("unchecked")
-    public static Iterable<Object> parseAll(String yaml) {
+    public static List<Object> parseAll(String yaml) {
```
Comment (ahgittin) [inline utils/common/src/main/java/org/apache/brooklyn/util/yaml/Yamls.java]: prefer the return type being `Iterable` as that allows lazy eval.  callers who need a list can do `MutableList.copyOf(Iterable)`.

File: launcher/pom.xml [hunk 2]
```diff
@@ -73,6 +73,11 @@
         </dependency>
         <dependency>
             <groupId>org.apache.brooklyn</groupId>
+            <artifactId>brooklyn-cm-salt</artifactId>
```
Comment (ahgittin) [inline launcher/pom.xml]: launcher should be quite bare.  i think the `all` project is where things like this are brought in.

**Comment by ahgittin** [comment]
The error looks like a problem with the jenkins/dev setup / assumption in where the UI is.  Probably better to declare a maven dependency and get it from the classpath. /cc @aledsage @ygy 

In any case the failure is unrelated.

```
2016-02-03 17:01:57,763 INFO  TESTNG INVOKING: "Surefire test" - org.apache.brooklyn.rest.jsgui.BrooklynJavascriptGuiLauncherTest.testJavascriptWithoutRest()
2016-02-03 17:01:57,768 WARN  Web application not found ../../brooklyn-ui/src/main/webapp
2016-02-03 17:01:57,774 WARN  Failed startup of context o.e.j.w.WebAppContext@6009cdc6{/,null,null}{../../brooklyn-ui/src/main/webapp}
java.io.FileNotFoundException: ../../brooklyn-ui/src/main/webapp
	at org.eclipse.jetty.webapp.WebInfConfiguration.unpack(WebInfConfiguration.java:493) ~[jetty-webapp-9.2.13.v20150730.jar:9.2.13.v20150730]
2016-02-03 17:01:57,775 INFO  JS GUI server started (no REST) at  http://localhost:8080/
```

**Comment by ahgittin** [comment]
@geomacy - i think this should go in `brooklyn-library` instead of `server` as it is optional.  `chef` and `winrm` should probably also be moved /cc @aledsage @hzbarcea .

probably also use package `o.a.b.entity.cm.salt` (add `entity`) wdyt?

otherwise - looks good at a high level - a couple minor comments i've made on lines of code. have **not** yet tested it though!

**Comment by grkvlt** [comment]
@ahgittin @aledsage @geomacy @hzbarcea Remember that there is that old SaltStack hackathon code in [brooklyn-library/sandbox/extra](https://github.com/apache/brooklyn-library/tree/master/sandbox/extra/src/main/java/org/apache/brooklyn/entity/salt) so when this gets moved there we should delete the old sandbox code, to prevent confusion.

**Comment by hzbarcea** [comment]
I am testing and reviewing now.
+1 on moving to `brooklyn-library`.
-0 on adding `entity`, don't see much value there.
+1 @grkvlt on removing the sandbox piece, obviously.

**Comment by geomacy** [comment]
If it helps for testing I have pushed the files I used to create a mysql formula and pillar data to a repository at https://github.com/geomacy/salt-ent-test.  I tarred these up as

```
$ tar tf my-mysql-formula.tar.gz
my-mysql-formula/
my-mysql-formula/mysql/
my-mysql-formula/mysql/files/
my-mysql-formula/mysql/files/classicmodels.schema
my-mysql-formula/mysql/files/foo.schema
```

and 
```
$ tar tf my-mysql-pillar.tar.gz
mysql/
mysql/init.sls
```

Then simply served them up from localhost with Node http-server.

The classicmodels.schema is from http://www.mysqltutorial.org.

**Comment by geomacy** [comment]
Closing this PR - as per the [comment above](https://github.com/apache/brooklyn-server/pull/4#issuecomment-179360438) I have opened a PR on the brooklyn-library repository instead, see https://github.com/apache/brooklyn-library/pull/1.

**Comment by geomacy** [comment]
Oops hit the wrong button- closing this again.

**Pull request closed without merging**